### PR TITLE
Custom type hints for all Pixeltable types

### PR DESCRIPTION
**Pull request by aaron-siegel**
Introduces custom type hints (`pxt.String`, `pxt.Array`, `pxt.Video`, etc.) for all Pixeltable types. The custom type hints are usable in both UDFs and schema definitions. UDFs can now be driven entirely by type hints, without the need for separate `param_types` and `return_type` clauses in the decorator.

Includes a custom mypy extension to ensure that the custom types are interpreted correctly during type checking.

### CODE REVIEW

**Comment by aaron-siegel** [comment]
This change is [Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/314)

**Comment by brucearctor** [comment]
Haven't really reviewed code, but the concept/description sounds like an awesome addition!

**Review by mkornacker — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/314)* status: 0 of 20 files reviewed, 6 unresolved discussions (waiting on @aaron-siegel)

___
*[`pixeltable/type_system.py` line 17 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/314#-O8cfH7V0lTCVU1wnVXx:-O8cfH7V0lTCVU1wnVXy:b19tfpz) ([raw file](https://github.com/pixeltable/pixeltable/blob/7e115780e31b81ab1a224c3b706a18d2d4d78b8c/pixeltable/type_system.py#L17)):*
> ```Python
> import PIL.Image
> import sqlalchemy as sql
> from typing_extensions import _AnnotatedAlias
> ```

Why do we need to use internals?
___
*[`pixeltable/type_system.py` line 262 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/314#-O8cfpzE2XCTSR3Co4_e:-O8cfpzE2XCTSR3Co4_f:b-bze7de) ([raw file](https://github.com/pixeltable/pixeltable/blob/7e115780e31b81ab1a224c3b706a18d2d4d78b8c/pixeltable/type_system.py#L262)):*
> ```Python
>         if typing.get_origin(t) is typing.Union:
>             union_args = typing.get_args(t)
>             if union_args[1] is type(None):
> ```

Is Union[None, T] also valid?
___
*[`pixeltable/type_system.py` line 899 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/314#-O8ceVR-4r-wCsa9989Q:-O8ceVR-4r-wCsa9989R:b-jv5vmb) ([raw file](https://github.com/pixeltable/pixeltable/blob/7e115780e31b81ab1a224c3b706a18d2d4d78b8c/pixeltable/type_system.py#L899)):*
> ```Python
> 
> 
> class _PxtType:
> ```

Missing class comment. In particular, the relationship to ColumnType and their respective responsibilities.
___
*[`pixeltable/type_system.py` line 954 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/314#-O8cgFT789ah2WdLUXRv:-O8cgFT789ah2WdLUXRw:bd9gzyf) ([raw file](https://github.com/pixeltable/pixeltable/blob/7e115780e31b81ab1a224c3b706a18d2d4d78b8c/pixeltable/type_system.py#L954)):*
> ```Python
> 
> class Array(np.ndarray, _PxtType):
>     def __class_getitem__(cls, item: Any) -> _AnnotatedAlias:
> ```

Function comment: what does item look like?
___
*[`pixeltable/type_system.py` line 982 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/314#-O8cgTHY0t_TvoHt0NHD:-O8cgTHY0t_TvoHt0NHE:buo9nmt) ([raw file](https://github.com/pixeltable/pixeltable/blob/7e115780e31b81ab1a224c3b706a18d2d4d78b8c/pixeltable/type_system.py#L982)):*
> ```Python
> 
> class Image(PIL.Image.Image, _PxtType):
>     def __class_getitem__(cls, item: Any) -> _AnnotatedAlias:
> ```

Same here
___
*[`tests/test_types.py` line 70 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/314#-O8chqAs6sj0kh31lcFR:-O8chqAs6sj0kh31lcFS:b-sqscx3) ([raw file](https://github.com/pixeltable/pixeltable/blob/7e115780e31b81ab1a224c3b706a18d2d4d78b8c/tests/test_types.py#L70)):*
> ```Python
>             assert ColumnType.from_python_type(Optional[py_type]) == pxt_type.copy(nullable=True)
> 
>         # Pixeltable types that are nullable by default and use T[NotNull] syntax
> ```

It doesn't use that syntax.

**Review by aaron-siegel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/314)* status: 0 of 20 files reviewed, 6 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/type_system.py` line 17 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/314#-O8cfH7V0lTCVU1wnVXx:-O8hBAKw2O4xJvWwZwM4:bj4h0g6) ([raw file](https://github.com/pixeltable/pixeltable/blob/7e115780e31b81ab1a224c3b706a18d2d4d78b8c/pixeltable/type_system.py#L17)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Why do we need to use internals?
</blockquote></details>

There's no other way that I'm aware of to get a type hint that represents a `typing.Annotated` instance. It's not actually a `type`.
___
*[`pixeltable/type_system.py` line 262 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/314#-O8cfpzE2XCTSR3Co4_e:-O8hBWj85hhmUWpmAA7u:bdgv1c4) ([raw file](https://github.com/pixeltable/pixeltable/blob/7e115780e31b81ab1a224c3b706a18d2d4d78b8c/pixeltable/type_system.py#L262)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Is Union[None, T] also valid?
</blockquote></details>

i guess, yeah, although I've never seen it used. But I suppose we should support it; I'll make the change
___
*[`pixeltable/type_system.py` line 899 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/314#-O8ceVR-4r-wCsa9989Q:-O8hGiO84Voj8gkdxZOU:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/7e115780e31b81ab1a224c3b706a18d2d4d78b8c/pixeltable/type_system.py#L899)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Missing class comment. In particular, the relationship to ColumnType and their respective responsibilities.
</blockquote></details>

Done.
___
*[`pixeltable/type_system.py` line 954 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/314#-O8cgFT789ah2WdLUXRv:-O8hHpvF9F_4bx3Osoib:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/7e115780e31b81ab1a224c3b706a18d2d4d78b8c/pixeltable/type_system.py#L954)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Function comment: what does item look like?
</blockquote></details>

Done.
___
*[`pixeltable/type_system.py` line 982 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/314#-O8cgTHY0t_TvoHt0NHD:-O8hHqWX5-gwhrXhLN0k:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/7e115780e31b81ab1a224c3b706a18d2d4d78b8c/pixeltable/type_system.py#L982)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Same here
</blockquote></details>

Done.
___
*[`tests/test_types.py` line 70 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/314#-O8chqAs6sj0kh31lcFR:-O8hI0Ww1YydkKb6Ft9H:bqljpdb) ([raw file](https://github.com/pixeltable/pixeltable/blob/7e115780e31b81ab1a224c3b706a18d2d4d78b8c/tests/test_types.py#L70)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

It doesn't use that syntax.
</blockquote></details>

Yes they do? I changed `use` to `accept` to clarify?

File: pixeltable/type_system.py [hunk 1]
```diff
@@ -897,6 +897,17 @@ def validate_media(self, val: Any) -> None:
...
 class _PxtType:
+    """
+    Base class for the Pixeltable type-hint family. Subclasses of this class are meant to be used as type hints, both
+    in schema definitions and in UDF signatures. Whereas `ColumnType`s are instantiable and carry semantic information
+    about the Pixeltable type system, `_PxtType` subclasses are purely for convenience: they are not instantiable and
+    must be resolved to a `ColumnType` (by calling `ColumnType.from_python_type()`) in order to do anything meaningful
+    with them.
+
+    `_PxtType` subclasses can be specialized (as type hints) with type parameters; for example:
+    `Image[(300, 300), 'RGB']`. The specialized forms resolve to `typing.Annotated` instances whose annotation is a
```
Comment (mkornacker) [inline pixeltable/type_system.py:938]: Should all these types be a typing.Annotated with an attached ColumnType, for the sake of uniformity?

**Review by aaron-siegel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/314)* status: 0 of 20 files reviewed, 7 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/type_system.py` line 908 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/314#gh-1793766462:-O8xPCZ4CSwe7Iz9mw2G:br3e8so) ([raw file](https://github.com/pixeltable/pixeltable/blob/ffa406e63e0f974b9a458e91fb13ac4fafd797ef/pixeltable/type_system.py#L908)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Should all these types be a typing.Annotated with an attached ColumnType, for the sake of uniformity? 
</blockquote></details>

cf. our earlier conversation, it's not possible for all of them to be typing.Annotated, since the non-scalar types need to be subscriptable.

Following the change to use Required[T] instead of T[NotNull], I did make the scalar types Annotated, which has some benefits.

**Review by mkornacker — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/314)* status: 0 of 20 files reviewed, 8 unresolved discussions (waiting on @aaron-siegel)

___
*[`pixeltable/type_system.py` line 168 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/314#-O8xozC58xwhkifdQ54B:-O8xozC58xwhkifdQ54C:b-424yyt) ([raw file](https://github.com/pixeltable/pixeltable/blob/405cc46ec86a70ecfc0e1d88c20967c3e8b65e92/pixeltable/type_system.py#L168)):*
> ```Python
> 
>     def __str__(self) -> str:
>         return self._type.name.lower()
> ```

Let's output the new syntax here:
String
or
Optional[String] (or: String | None)

**Review by aaron-siegel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/314)* status: 0 of 52 files reviewed, 8 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/type_system.py` line 168 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/314#-O8xozC58xwhkifdQ54B:-O963LtOBZ2IqvG5BVgp:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/405cc46ec86a70ecfc0e1d88c20967c3e8b65e92/pixeltable/type_system.py#L168)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Let's output the new syntax here:
String
or
Optional[String] (or: String | None)
</blockquote></details>

Done.

**Review by aaron-siegel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/314)* status: 0 of 52 files reviewed, 16 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/type_system.py` line 914 at r4](https://reviewable.io/reviews/pixeltable/pixeltable/314#-O9Hr3vU9D7i-HN_7fMb:-O9JHcPN3QFk2f2PoexS:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/0eca345cf55e2b7b2415fd852acc6270dc86520c/pixeltable/type_system.py#L914)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Add explanatory comment
</blockquote></details>

Done.
___
*[`tests/test_function.py` line 314 at r4](https://reviewable.io/reviews/pixeltable/pixeltable/314#-O9HsRw2C21AxXeoQnqt:-O9JHisF4YWtmj0ehk_b:bg9it75) ([raw file](https://github.com/pixeltable/pixeltable/blob/f6a3d82c53d49f115c0d74ad1f77f3e0498e3720/tests/test_function.py#L314)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

These are required.
</blockquote></details>

Ok, I gather you want to see a `pxt.Required` any place there was a `nullable=False` previously. Let's discuss at standup; I don't understand why it's important here.

I'd presumed it doesn't really matter since the tests pass either way, so marking them as Required or not doesn't actually have a coverage impact in these cases. I was defaulting to "use the pattern that a user would be most likely to use in a similar circumstance", which is to not use Required unless there's a specific reason to do so.

I don't feel super strongly about it, but I'd like to understand the rationale before I go and make the change.
___
*[`tests/test_table.py` line 210 at r4](https://reviewable.io/reviews/pixeltable/pixeltable/314#-O9Ht9ml-ubvVQOeBICc:-O9JJ12D0LpiUjemwDXh:b-wcnodm) ([raw file](https://github.com/pixeltable/pixeltable/blob/f6a3d82c53d49f115c0d74ad1f77f3e0498e3720/tests/test_table.py#L210)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

There are a bunch of places where nullable=False is missing, let's keep those semantics.
</blockquote></details>

As above - why?
___
*[`tests/test_types.py` line 47 at r4](https://reviewable.io/reviews/pixeltable/pixeltable/314#-O9HthKJ7MOETj2nu98C:-O9JJCs0CRAS9nUcKe6s:bt2vr1b) ([raw file](https://github.com/pixeltable/pixeltable/blob/0eca345cf55e2b7b2415fd852acc6270dc86520c/tests/test_types.py#L47)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

The type decl is wrong
</blockquote></details>

Fixed.
___
*[`tests/test_types.py` line 49 at r4](https://reviewable.io/reviews/pixeltable/pixeltable/314#-O9HuD3L8nCCnpmyy888:-O9JJQyU9m5jDUlF5Srg:bxttxog) ([raw file](https://github.com/pixeltable/pixeltable/blob/0eca345cf55e2b7b2415fd852acc6270dc86520c/tests/test_types.py#L49)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Why not also test Optional[str]?
</blockquote></details>

`Optional[str]` is included; see code below. All of these cases get tested in each of their default (`T`), optional (`Optional[T]`), and required (`Required[T]`) forms.

**Review by mkornacker — APPROVED** [review]

**Pull request merged**